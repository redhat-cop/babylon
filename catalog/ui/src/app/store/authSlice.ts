import {
  createAction,
  createReducer,
} from '@reduxjs/toolkit';

import {
  createSelector,
} from 'reselect'

// reducer logic

function reduce_clearImpersonateUser(authState, action) {
  authState.impersonateUser = null;
  authState.catalogNamespaces = authState.authCatalogNamespaces;
  authState.userNamespace = authState.authUserNamespace;
}

function reduce_setImpersonateUser(authState, action) {
  const {user, catalogNamespaces, userNamespace} = action.payload;
  authState.impersonateUser = user;
  authState.catalogNamespaces = catalogNamespaces;
  authState.userNamespace = userNamespace;
}

function reduce_startSession(authState, action) {
  const {admin, user, userNamespace, catalogNamespaces} = action.payload;
  authState.admin = admin;
  authState.authUser = user;
  authState.authUserNamespace = userNamespace;
  authState.authCatalogNamespaces = catalogNamespaces;
  authState.userNamespace = userNamespace;
  authState.catalogNamespaces = catalogNamespaces;
}

export const authReducer = createReducer({}, {
  "auth/clearImpersonateUser": reduce_clearImpersonateUser,
  "auth/setImpersonateUser": reduce_setImpersonateUser,
  "auth/startSession": reduce_startSession,
});


// Action creators
export const actionClearImpersonateUser = createAction("auth/clearImpersonateUser");
export const actionSetImpersonateUser = createAction("auth/setImpersonateUser");
export const actionStartSession = createAction("auth/startSession");


// Selectors
export const selectActiveUser = createSelector(
  state => state.auth,
  auth => auth.impersonateUser || auth.authUser,
)

export const selectAuthUser = createSelector(
  state => state.auth,
  auth => auth.authUser,
)

export const selectImpersonateUser = createSelector(
  state => state.auth,
  auth => auth.impersonateUser,
)

export const selectUserNamespace = createSelector(
  state => state.auth,
  auth => auth.userNamespace,
)

export const selectCatalogNamespaces = createSelector(
  state => state.auth,
  auth => auth.catalogNamespaces,
)
