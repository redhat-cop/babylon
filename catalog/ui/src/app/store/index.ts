import { createAction, createReducer, configureStore } from '@reduxjs/toolkit';
import { createSelector } from 'reselect';
import { CatalogNamespace, ServiceNamespace } from '@app/types';

export interface ActionSetImpersonation {
  admin: string;
  user: string;
  groups: string[];
  catalogNamespaces: [];
  serviceNamespaces: [];
  userNamespace: [];
}

export interface ActionStartSession {
  admin: boolean;
  consoleURL: string;
  user: string;
  groups: string[];
  roles: string[];
  interface: string;
  catalogNamespaces: CatalogNamespace[];
  serviceNamespaces: ServiceNamespace[];
  userNamespace: ServiceNamespace;
}

// Reducer functions
function reduce_clearImpersonation(state) {
  state.impersonate = null;
  state.catalogItems = null;
  sessionStorage.removeItem('impersonateUser');
}
function reduce_setImpersonation(state, action) {
  const { admin, groups, roles, user, catalogNamespaces, serviceNamespaces, userNamespace } = action.payload;
  state.impersonate = {
    admin: admin || false,
    groups: groups || [],
    roles: roles || [],
    user: user,
    catalogNamespaces: catalogNamespaces,
    serviceNamespaces: serviceNamespaces,
    userNamespace: userNamespace,
  };
  state.catalogItems = null;
  state.resourceClaims = null;
}

function reduce_startSession(state, action) {
  state.auth.admin = action.payload.admin || false;
  state.auth.groups = action.payload.groups || [];
  state.auth.roles = action.payload.roles || [];
  state.auth.user = action.payload.user;
  state.auth.catalogNamespaces = action.payload.catalogNamespaces;
  state.auth.serviceNamespaces = action.payload.serviceNamespaces;
  state.auth.userNamespace = action.payload.userNamespace;
  state.catalogItems = null;
  state.consoleURL = action.payload.consoleURL;
  state.interface = action.payload.interface;
}

// Action creators
export const actionClearImpersonation = createAction('clearImpersonation');
export const actionSetImpersonation = createAction<ActionSetImpersonation>('setImpersonation');
export const actionStartSession = createAction<ActionStartSession>('startSession');

// Selectors
const selectSelf = (state: any) => state;
const selectAuth = (state: any) => (state.impersonate !== null ? state.impersonate : state.auth);

export const selectAuthIsAdmin = createSelector(selectSelf, (state: any): boolean => state.auth.admin);

export const selectAuthUser = createSelector(selectSelf, (state: any): string => state.auth.user);

export const selectConsoleURL = createSelector(selectSelf, (state: any): string => state.consoleURL);

export const selectInterface = createSelector(selectSelf, (state: any): string => state.interface);

export const selectUser = createSelector(selectAuth, (state: any): string => state.user);

export const selectUserGroups = createSelector(selectAuth, (state: any): string[] =>
  (state.groups || []).filter(Boolean).concat('system:authenticated'),
);

export const selectUserIsAdmin = createSelector(selectAuth, (state: any): boolean => state.admin);

export const selectUserRoles = createSelector(selectAuth, (state: any): string[] =>
  (state.roles || []).filter(Boolean),
);

export const selectImpersonationUser = createSelector(selectSelf, (state: any): string => state.impersonate?.user);

export const selectCatalogNamespace = createSelector(
  [(state: any) => selectCatalogNamespaces(state), (state: any, namespace: string): string => namespace],
  (catalogNamespaces: any, namespace: string): CatalogNamespace =>
    (catalogNamespaces || []).filter(Boolean).find((catalogNamespace) => catalogNamespace.name === namespace),
);

export const selectCatalogNamespaces = createSelector(selectAuth, (state: any): CatalogNamespace[] =>
  (state.catalogNamespaces || []).filter(Boolean),
);

export const selectServiceNamespace = createSelector(
  [(state: any) => state.impersonate || state.auth, (state: any, namespace: string): string => namespace],
  (state: any, namespace: string) => (state.serviceNamespaces || []).filter(Boolean).find((ns) => ns.name == namespace),
);

export const selectServiceNamespaces = createSelector(
  (state: any) => state.impersonate || state.auth,
  (state: any) => (state.serviceNamespaces || []).filter(Boolean),
);

export const selectUserNamespace = createSelector(selectAuth, (state) => state.userNamespace);

// Store
export const store = configureStore({
  reducer: createReducer(
    {
      auth: {
        admin: null,
        groups: [],
        user: null,
        catalogNamespaces: [],
        serviceNamespaces: [],
        userNamespace: null,
      },
      catalogItems: null,
      impersonate: null,
      interface: null,
    },
    {
      clearImpersonation: reduce_clearImpersonation,
      setImpersonation: reduce_setImpersonation,
      startSession: reduce_startSession,
    },
  ),
});
