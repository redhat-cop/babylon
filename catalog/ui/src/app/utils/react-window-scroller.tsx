import { useRef, useEffect, useCallback } from 'react';
import throttle from 'lodash.throttle';

const windowScrollPositionKey = {
  y: 'pageYOffset',
  x: 'pageXOffset',
};

const documentScrollPositionKey = {
  y: 'scrollTop',
  x: 'scrollLeft',
};

const getScrollPosition = (axis: 'x' | 'y'): number =>
  window[windowScrollPositionKey[axis]] ||
  document.documentElement[documentScrollPositionKey[axis]] ||
  document.body[documentScrollPositionKey[axis]] ||
  0;

export const ReactWindowScroller = ({ children, throttleTime = 10, isGrid = false }): JSX.Element => {
  const ref = useRef();
  const outerRef = useRef();

  useEffect(() => {
    const handleWindowScroll = throttle(() => {
      const { offsetTop, offsetLeft } = outerRef.current || { offsetLeft: 0, offsetTop: 0 };
      const scrollTop = getScrollPosition('y') - offsetTop;
      const scrollLeft = getScrollPosition('x') - offsetLeft;
      // @ts-ignore: Object is possibly 'null'.
      if (isGrid && ref.current) ref.current.scrollTo({ scrollLeft, scrollTop });
      // @ts-ignore: Object is possibly 'null'.
      if (!isGrid && ref.current) ref.current.scrollTo(scrollTop);
    }, throttleTime);

    window.addEventListener('scroll', handleWindowScroll);
    return () => {
      handleWindowScroll.cancel();
      window.removeEventListener('scroll', handleWindowScroll);
    };
  }, [isGrid]);

  const onScroll = useCallback(
    ({ scrollLeft, scrollTop, scrollOffset, scrollUpdateWasRequested }) => {
      if (!scrollUpdateWasRequested) return;
      const top = getScrollPosition('y');
      const left = getScrollPosition('x');
      const { offsetTop, offsetLeft } = outerRef.current || { offsetLeft: 0, offsetTop: 0 };

      scrollOffset += Math.min(top, offsetTop);
      scrollTop += Math.min(top, offsetTop);
      scrollLeft += Math.min(left, offsetLeft);

      if (!isGrid && scrollOffset !== top) window.scrollTo(0, scrollOffset);
      if (isGrid && (scrollTop !== top || scrollLeft !== left)) {
        window.scrollTo(scrollLeft, scrollTop);
      }
    },
    [isGrid]
  );

  return children({
    ref,
    outerRef,
    style: {
      width: isGrid ? 'auto' : '100%',
      height: '100%',
      display: 'inline-block',
    },
    onScroll,
  });
};
