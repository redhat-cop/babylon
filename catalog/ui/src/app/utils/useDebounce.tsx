import { useEffect, useRef, useState } from 'react';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function useDebounce(inner: (...args: any[]) => unknown, ms = 0): (...args: unknown[]) => Promise<unknown> {
  const [timer, setTimer] = useState(null);
  const [resolves, setResolves] = useState([]);
  const callbackRef = useRef((...args: unknown[]) => {
    clearTimeout(timer);
    setTimer(
      setTimeout(() => {
        // Get the result of the inner function, then apply it to the resolve function of
        // each promise that has been created since the last time the inner function was run
        const result = inner(...args);
        resolves.forEach((r) => r(result));
        setResolves([]);
      }, ms),
    );

    return new Promise((r) => resolves.push(r));
  });
  useEffect(() => {
    return () => clearTimeout(timer);
  }, [timer]);

  return callbackRef.current;
}

export default useDebounce;
