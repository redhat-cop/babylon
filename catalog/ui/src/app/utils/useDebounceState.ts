import { useState, useEffect } from 'react';

function useDebounceState(value: string, delay: number) {
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);

    return () => {
      clearTimeout(handler);
    };
  }, [value, delay]);

  return debouncedValue;
}

export default useDebounceState;
