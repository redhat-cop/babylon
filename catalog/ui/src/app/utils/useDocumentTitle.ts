import { useEffect } from 'react';
import { useSearchParams } from 'react-router-dom';

const DEFAULT_USER_INTERFACE = 'rhpds';

function getPageTitle(title: string, userInterface?: string): string {
  if (!userInterface) userInterface = DEFAULT_USER_INTERFACE;
  return userInterface === 'summit'
    ? title.replace('Babylon', 'Red Hat Summit')
    : ['rhpds', 'rhdp', 'rhdp-partners'].includes(userInterface)
      ? title.replace('Babylon', 'Red Hat Demo Platform')
      : title;
}

// a custom hook for setting the page title
function useDocumentTitle(title: string): void {
  const [searchParams] = useSearchParams();
  const userInterface = searchParams.get('userInterface');
  useEffect(() => {
    const originalTitle = document.title;
    if (title) document.title = getPageTitle(title, userInterface);

    return () => {
      document.title = originalTitle;
    };
  }, [title, userInterface]);
}

export default useDocumentTitle;
