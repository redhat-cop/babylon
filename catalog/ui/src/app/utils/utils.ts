/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable @typescript-eslint/no-explicit-any */

import { K8sObject } from '@app/types';

export function accessibleRouteChangeHandler() {
  return window.setTimeout(() => {
    const mainContainer = document.getElementById('primary-app-container');
    if (mainContainer) {
      mainContainer.focus();
    }
  }, 50);
}

export function k8sObjectListEquals(newObjList: K8sObject[], oldObjList: K8sObject[]) {
  for (const obj of newObjList) {
    let itemEqual = false;
    oldObjList.forEach((c: any) => {
      if (c.metadata.uid === obj.metadata.uid) {
        itemEqual = c.metadata.resourceVersion === obj.metadata.resourceVersion;
      }
    });
    if (!itemEqual) {
      return false;
    }
  }
  return true;
}
