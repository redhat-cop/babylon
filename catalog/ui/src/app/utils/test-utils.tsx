import React, { ReactElement } from 'react';
import { render, RenderOptions, queries, RenderResult } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '@app/store';
import { createMemoryHistory, MemoryHistory } from 'history';

const AllTheProviders = ({ children, history }) => {
  return (
    <Provider store={store}>
      <Router history={history}>{children}</Router>
      <div id="modal-root"></div>
    </Provider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: { history: MemoryHistory<unknown> } & Omit<RenderOptions, 'queries'>
): RenderResult<typeof queries> => {
  function getOptions({ history = createMemoryHistory(), ...rest }) {
    return { rest, history };
  }
  const { history, ...rest } = getOptions(options || {});
  return render(ui, {
    wrapper: ({ children }) => <AllTheProviders history={history}>{children}</AllTheProviders>,
    ...rest,
  });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
