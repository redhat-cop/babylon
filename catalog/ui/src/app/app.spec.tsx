import React from 'react';
import App from '@app/index';
import { render } from '@testing-library/react';
import { fireEvent, generateSession } from './utils/test-utils';
import { Provider } from 'react-redux';
import { store } from './store';

jest.mock('@app/utils/useSession', () =>
  jest.fn(() => ({
    getSession: () => generateSession({}),
  })),
);

const wrapper = (children: React.ReactNode) => (
  <Provider store={store}>
    {children}
    <div id="modal-root"></div>
  </Provider>
);

describe('App tests', () => {
  it.only('should render a nav-toggle button', () => {
    const { container } = render(wrapper(<App />));
    const button = container.querySelector('#nav-toggle');
    expect(button).toBeInTheDocument();
  });

  it('should hide the sidebar on smaller viewports', () => {
    Object.defineProperty(window, 'innerWidth', { writable: true, configurable: true, value: 600 });
    const { container } = render(wrapper(<App />));
    fireEvent(window, new Event('resize'));
    expect(container.querySelector('#page-sidebar').classList.contains('pf-m-collapsed')).toBeTruthy();
  });

  it('should expand the sidebar on larger viewports', () => {
    Object.defineProperty(window, 'innerWidth', { writable: true, configurable: true, value: 1200 });
    const { container } = render(wrapper(<App />));
    fireEvent.resize(window);
    expect(container.querySelector('#page-sidebar').classList.contains('pf-m-expanded')).toBeTruthy();
  });

  it('should hide the sidebar when clicking the nav-toggle button', async () => {
    Object.defineProperty(window, 'innerWidth', { writable: true, configurable: true, value: 1200 });
    const { container } = render(wrapper(<App />));
    fireEvent(window, new Event('resize'));
    const button = container.querySelector('#nav-toggle');
    expect(container.querySelector('#page-sidebar').classList.contains('pf-m-expanded')).toBeTruthy();
    fireEvent.click(button);
    expect(container.querySelector('#page-sidebar').classList.contains('pf-m-collapsed')).toBeTruthy();
    expect(container.querySelector('#page-sidebar').classList.contains('pf-m-expanded')).toBeFalsy();
  });
});
