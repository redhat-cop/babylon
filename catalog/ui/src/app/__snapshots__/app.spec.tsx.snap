// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests should render default App component 1`] = `
<BrowserRouter>
  <Switch>
    <Route
      path="/workshop"
    >
      <Workshop />
    </Route>
    <Route
      path="/"
    >
      <AppLayout>
        <AppRoutes />
      </AppLayout>
    </Route>
  </Switch>
</BrowserRouter>
`;
